### PR TITLE
[#62] [Android] [UI] As a user, I can see text area answer

### LINKED ISSUE #62: [Android] [UI] As a user, I can see text area answer
## Why
A text-area question type is a question type that allows a user to fill in a multi-line text input. It is commonly used when we would like to ask an open-end question, such as asking for suggestions or comments.

### Example questions
- Do you have any suggestions?
- Your comments are appreciated!

## Acceptance Criteria
For text-area questions,
- Display a textbox that supports multi-line text input
    - Display a placeholder inside the textbox
    - The textbox is scrollable when the number of lines goes beyond its viewport (i.e., when the text input no longer fits within the textbox)
- Display the keyboard when a textbox is selected
    - Tap anywhere outside of textbox to release the keyboard
    - The keyboard must not cover what the user is typing

## Design

[Text-area Question](https://www.figma.com/file/GjRPOjDyZ6f4EDL3wKarRK/Challenge-Mobile-App?node-id=2632%3A674)

### PR DESCRIPTION
**Pull request by suho**
- Resolves #62

## What happened 👀

Added the text area answer for the survey questions

## Insight 📝

- A minor update: I changed from `onAnswerClick` to `onAnswerChange` for more general.
- I also added a new data class: AnswerInput to contain answer information

## Proof Of Work 📹

Screenshot 2022-12-14 at 16 43 17

### CODE REVIEW

**Review by hoangnguyen92dn — APPROVED** [review]
LGTM 🚀

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/domain/model/Question.kt [hunk 1]
```diff
@@ -49,7 +51,9 @@ fun Question.displayType(): QuestionDisplayType {
             it.text ?: ""
         })
         QUESTION_DISPLAY_TYPE_NPS -> QuestionDisplayType.Nps
-        QUESTION_DISPLAY_TYPE_TEXTAREA -> QuestionDisplayType.Textarea
+        QUESTION_DISPLAY_TYPE_TEXTAREA -> QuestionDisplayType.Textarea(
+            sortedAnswers().first().inputMaskPlaceholder ?: ""
```
[the remark this answers is not in the document]
Comment (suho) [reply shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/domain/model/Question.kt]: Fixed in https://github.com/suho/kmm-ic/pull/119/commits/0eccb3deb4d926fd74cf61595ea6642f8cdba07e

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt [hunk 2]
```diff
@@ -61,15 +61,15 @@ fun SurveyQuestionsScreen(
...
+        onAnswerChange = viewModel::answerQuestion
     )
 }
 
 @Composable
 private fun SurveyQuestionsScreenContent(
     uiModel: SurveyQuestionsContentUiModel,
     onCloseClick: () -> Unit = {},
-    onAnswerClick: (Pair<Int, List<Int>>) -> Unit
+    onAnswerChange: (Pair<Int, List<AnswerInput>>) -> Unit
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt:72]: Should we define the DTO for the whole `Pair<Int, List<AnswerInput>>` data structure?

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt [hunk 3]
```diff
@@ -257,18 +257,25 @@ private fun Answer(type: QuestionDisplayType, onAnswersClick: (List<Int>) -> Uni
...
         is QuestionDisplayType.Choice -> MultipleChoicesAnswer(
             choices = type.answers,
             onIndexesChange = {
-                onAnswersClick(it.toList())
+                onAnswersChange(it.toList().map(AnswerInput::Index))
             }
         )
+        is QuestionDisplayType.Textarea -> TextareaAnswer(
+            onTextChange = {
+                onAnswersChange(listOf(AnswerInput.Content(index = 0, content = it)))
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt:274]: Should the `Content` data class need an index field?

File: shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/domain/model/Question.kt [hunk 4]
```diff
@@ -49,7 +51,9 @@ fun Question.displayType(): QuestionDisplayType {
             it.text ?: ""
         })
         QUESTION_DISPLAY_TYPE_NPS -> QuestionDisplayType.Nps
-        QUESTION_DISPLAY_TYPE_TEXTAREA -> QuestionDisplayType.Textarea
+        QUESTION_DISPLAY_TYPE_TEXTAREA -> QuestionDisplayType.Textarea(
+            sortedAnswers().first().inputMaskPlaceholder.orEmpty()
+        )
```
Comment (luongvo) [inline shared/src/commonMain/kotlin/co/nimblehq/ic/kmm/suv/domain/model/Question.kt:56]: Following my concern [here](https://github.com/suho/kmm-ic/pull/118#discussion_r1055069609), we can get the `first().inputMaskPlaceholder.orEmpty()` from `QuestionContentUiModel.answers` at SurveyQuestionsScreen. 🤔

**Review by luongvo — COMMENTED** [review]
rest lgtm

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt (hunk 3 above)
Comment (luongvo) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt:274]: Ok, we need it 😄 https://github.com/suho/kmm-ic/pull/120/files#diff-41fd63ac41b8847df57e438ab9f3c99dcb8353c1adc4bb57302d8618307ed8efR279-R283

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt (hunk 2 above)
Comment (suho) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsScreen.kt]: Fixed in https://github.com/suho/kmm-ic/pull/119/commits/a32b80000213f9fff8004060a41930e1bb732233

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt [hunk 5]
```diff
@@ -53,7 +53,9 @@ class SurveyQuestionsViewModel(
         }
     }
 
-    fun answerQuestion(questionAnswer: Pair<Int, List<Int>>) {
-        Timber.d("question ${questionAnswer.first} - answers ${questionAnswer.second}")
+    fun answerQuestion(questionAndAnswer: QuestionAndAnswers) {
```
Comment (luongvo) [inline android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt:56]: mismatching param name 😂 

```suggestion
    fun answerQuestion(questionAndAnswers: QuestionAndAnswers) {
```

**Review by luongvo — COMMENTED** [review]
1 minor issue @suho 🧑‍🍳

**Comment by luongvo** [comment]
CI fails 🤓

File: android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt (hunk 5 above)
Comment (suho) [reply android/src/main/java/co/nimblehq/ic/kmm/suv/android/ui/screens/surveyquestions/SurveyQuestionsViewModel.kt]: Fixed in https://github.com/suho/kmm-ic/pull/119/commits/7807069f58e6f82988f546500774719b119e6fae

**Review by luongvo — APPROVED** [review]

**Pull request merged**